File: features/health/data/risk_engine.dart
import 'dart:convert';
import 'package:crypto/crypto.dart';
import 'package:flutter/foundation.dart';
import 'package:flock_sense/features/daily_records/domain/daily_record_model.dart';
import 'package:flock_sense/features/health/data/risk_config.dart';
import 'package:flock_sense/features/health/domain/health_case_model.dart';
import 'package:flock_sense/features/health/domain/risk_assessment_model.dart';
import 'package:flock_sense/features/vaccine/domain/vaccine_record_model.dart';

/// Calculation Input Payload for Risk Engine
class RiskEngineInput {
  final String caseId;
  final String farmId;
  final String batchId;
  final int currentMortality;
  final int affectedCount;
  final List<String> symptoms;
  final double? feedReductionPercent;
  final double? waterReductionPercent;
  final double? temperature;
  final double? humidity;
  final DateTime incidentDate;
  final int currentBirdCount;
  final List<DailyRecordModel> recentDailyRecords;
  final List<VaccineRecordModel> vaccinationRecords;
  final List<HealthCaseModel> recentHealthCases;
  final bool hasOverdueVaccine;

  const RiskEngineInput({
    required this.caseId,
    required this.farmId,
    required this.batchId,
    required this.currentMortality,
    required this.affectedCount,
    required this.symptoms,
    this.feedReductionPercent,
    this.waterReductionPercent,
    this.temperature,
    this.humidity,
    required this.incidentDate,
    this.currentBirdCount = 1000,
    this.recentDailyRecords = const [],
    this.vaccinationRecords = const [],
    this.recentHealthCases = const [],
    this.hasOverdueVaccine = false,
  });

  /// Deterministic Hash for Trigger-Loop Prevention & Idempotency
  String computeInputHash() {
    final normalizedSymptoms = List<String>.from(symptoms)..sort();
    final raw = [
      caseId,
      farmId,
      batchId,
      currentMortality,
      affectedCount,
      normalizedSymptoms.join(','),
      feedReductionPercent?.toStringAsFixed(1) ?? '0',
      waterReductionPercent?.toStringAsFixed(1) ?? '0',
      temperature?.toStringAsFixed(1) ?? '0',
      humidity?.toStringAsFixed(1) ?? '0',
      hasOverdueVaccine ? '1' : '0',
      RiskConfig.engineVersion,
    ].join('|');

    return md5.convert(utf8.encode(raw)).toString();
  }
}

/// Centralized Explainable Health Risk Engine (SIH26128)
class RiskEngine {
  RiskEngine._();

  /// Primary pure calculation method
  static RiskAssessmentModel calculateHealthRisk(RiskEngineInput input) {
    final structuredReasons = <RiskReasonModel>[];
    final symptomSignals = <String>[];
    final vaccinationSignals = <String>[];
    final environmentalSignals = <String>[];
    var rawScore = 0;

    // ── 1. Mortality Baseline & Spike Analysis (Max 25 pts) ───────────────────
    final baselineResult = _analyzeMortality(
      currentMortality: input.currentMortality,
      currentBirdCount: input.currentBirdCount,
      recentRecords: input.recentDailyRecords,
      incidentDate: input.incidentDate,
    );

    rawScore += baselineResult.pointsAdded;
    if (baselineResult.reason != null) {
      structuredReasons.add(baselineResult.reason!);
    }

    // ── 2. Feed Reduction Analysis (Max 15 pts) ──────────────────────────────
    final feedReduction = input.feedReductionPercent ?? 0.0;
    if (feedReduction > RiskConfig.feedReductionCritical) {
      const pts = RiskConfig.feedPointsCritical;
      rawScore += pts;
      structuredReasons.add(const RiskReasonModel(
        code: 'feed_reduction_critical',
        label: 'Severe feed consumption drop',
        value: '>25% intake loss',
        points: pts,
        severity: 'critical',
      ));
    } else if (feedReduction > RiskConfig.feedReductionHigh) {
      const pts = RiskConfig.feedPointsHigh;
      rawScore += pts;
      structuredReasons.add(RiskReasonModel(
        code: 'feed_reduction_high',
        label: 'Significant feed intake reduction',
        value: '-${feedReduction.toStringAsFixed(0)}%',
        points: pts,
        severity: 'warning',
      ));
    } else if (feedReduction > RiskConfig.feedReductionModerate) {
      const pts = RiskConfig.feedPointsModerate;
      rawScore += pts;
      structuredReasons.add(RiskReasonModel(
        code: 'feed_reduction_moderate',
        label: 'Moderate feed intake drop',
        value: '-${feedReduction.toStringAsFixed(0)}%',
        points: pts,
        severity: 'warning',
      ));
    } else if (feedReduction > RiskConfig.feedReductionMild) {
      const pts = RiskConfig.feedPointsMild;
      rawScore += pts;
      structuredReasons.add(RiskReasonModel(
        code: 'feed_reduction_mild',
        label: 'Minor feed intake reduction',
        value: '-${feedReduction.toStringAsFixed(0)}%',
        points: pts,
        severity: 'info',
      ));
    }

    // ── 3. Water Reduction Analysis (Max 15 pts) ─────────────────────────────
    final waterReduction = input.waterReductionPercent ?? 0.0;
    if (waterReduction > RiskConfig.waterReductionCritical) {
      const pts = RiskConfig.waterPointsCritical;
      rawScore += pts;
      structuredReasons.add(const RiskReasonModel(
        code: 'water_reduction_critical',
        label: 'Severe water consumption drop',
        value: '>25% water loss',
        points: pts,
        severity: 'critical',
      ));
    } else if (waterReduction > RiskConfig.waterReductionHigh) {
      const pts = RiskConfig.waterPointsHigh;
      rawScore += pts;
      structuredReasons.add(RiskReasonModel(
        code: 'water_reduction_high',
        label: 'Significant water intake drop',
        value: '-${waterReduction.toStringAsFixed(0)}%',
        points: pts,
        severity: 'warning',
      ));
    } else if (waterReduction > RiskConfig.waterReductionModerate) {
      const pts = RiskConfig.waterPointsModerate;
      rawScore += pts;
      structuredReasons.add(RiskReasonModel(
        code: 'water_reduction_moderate',
        label: 'Moderate water intake drop',
        value: '-${waterReduction.toStringAsFixed(0)}%',
        points: pts,
        severity: 'warning',
      ));
    } else if (waterReduction > RiskConfig.waterReductionMild) {
      const pts = RiskConfig.waterPointsMild;
      rawScore += pts;
      structuredReasons.add(RiskReasonModel(
        code: 'water_reduction_mild',
        label: 'Minor water intake reduction',
        value: '-${waterReduction.toStringAsFixed(0)}%',
        points: pts,
        severity: 'info',
      ));
    }

    // ── 4. Clinical Symptom Weighting (Capped at 25 pts) ─────────────────────
    var rawSymptomScore = 0;
    final normalizedSymptoms = input.symptoms.map((s) => s.trim().toLowerCase().replaceAll(' ', '_')).toList();

    for (final symptom in normalizedSymptoms) {
      final weight = RiskConfig.symptomWeights[symptom] ?? 4;
      rawSymptomScore += weight;
      symptomSignals.add(symptom);
    }

    final effectiveSymptomPoints = rawSymptomScore > RiskConfig.maxSymptomPoints
        ? RiskConfig.maxSymptomPoints
        : rawSymptomScore;

    if (effectiveSymptomPoints > 0) {
      rawScore += effectiveSymptomPoints;
      final severity = effectiveSymptomPoints >= 15
          ? 'critical'
          : (effectiveSymptomPoints >= 8 ? 'warning' : 'info');
      structuredReasons.add(RiskReasonModel(
        code: 'clinical_symptoms',
        label: 'Clinical symptoms reported',
        value: '${normalizedSymptoms.length} signs (${effectiveSymptomPoints} pts)',
        points: effectiveSymptomPoints,
        severity: severity,
      ));
    }

    // ── 5. Syndromic Cluster Bonuses (+5 pts each) ───────────────────────────
    final respMatches = normalizedSymptoms.where((s) => RiskConfig.respiratorySymptoms.contains(s)).length;
    if (respMatches >= 2) {
      const pts = RiskConfig.syndromeBonusPoints;
      rawScore += pts;
      structuredReasons.add(RiskReasonModel(
        code: 'respiratory_syndrome',
        label: 'Respiratory syndrome detected',
        value: '$respMatches distinct respiratory signs',
        points: pts,
        severity: 'critical',
      ));
    }

    final digestMatches = normalizedSymptoms.where((s) => RiskConfig.digestiveSymptoms.contains(s)).length;
    if (digestMatches >= 2) {
      const pts = RiskConfig.syndromeBonusPoints;
      rawScore += pts;
      structuredReasons.add(RiskReasonModel(
        code: 'digestive_syndrome',
        label: 'Digestive / enteric syndrome detected',
        value: '$digestMatches enteric signs',
        points: pts,
        severity: 'warning',
      ));
    }

    final neuroMatches = normalizedSymptoms.where((s) => RiskConfig.neurologicalSymptoms.contains(s)).length;
    if (neuroMatches >= 1) {
      const pts = RiskConfig.syndromeBonusPoints;
      rawScore += pts;
      structuredReasons.add(RiskReasonModel(
        code: 'neurological_syndrome',
        label: 'Neurological distress signs detected',
        value: '$neuroMatches neuro sign',
        points: pts,
        severity: 'critical',
      ));
    }

    // ── 6. Sudden Death Acute Signal (+10 pts) ────────────────────────────────
    if (normalizedSymptoms.contains('sudden_death')) {
      const pts = RiskConfig.suddenDeathBonusPoints;
      rawScore += pts;
      structuredReasons.add(const RiskReasonModel(
        code: 'sudden_death_event',
        label: 'Sudden acute mortality event reported',
        value: 'Peracute onset',
        points: pts,
        severity: 'critical',
      ));
    }

    // ── 7. Vaccination Compliance Signal (Max 10 pts) ─────────────────────────
    if (input.hasOverdueVaccine) {
      const pts = RiskConfig.vaccineSignificantlyOverduePoints;
      rawScore += pts;
      vaccinationSignals.add('overdue');
      structuredReasons.add(const RiskReasonModel(
        code: 'vaccine_overdue',
        label: 'Flock immunization overdue',
        value: 'Missing scheduled vaccine',
        points: pts,
        severity: 'warning',
      ));
    } else if (input.vaccinationRecords.isEmpty) {
      const pts = RiskConfig.vaccineUnknownPoints;
      rawScore += pts;
      vaccinationSignals.add('unknown');
      structuredReasons.add(const RiskReasonModel(
        code: 'vaccine_unverified',
        label: 'Vaccination records unrecorded',
        value: 'Incomplete immunization log',
        points: pts,
        severity: 'info',
      ));
    } else {
      vaccinationSignals.add('up_to_date');
    }

    // ── 8. Environmental Stress Signal (Max 10 pts) ───────────────────────────
    var envPoints = 0;
    if (input.temperature != null) {
      final t = input.temperature!;
      if (t < RiskConfig.optimalTempMin || t > RiskConfig.optimalTempMax) {
        envPoints += RiskConfig.envStressModeratePoints;
        environmentalSignals.add('temp_out_of_range');
      }
    }
    if (input.humidity != null) {
      final h = input.humidity!;
      if (h < RiskConfig.optimalHumidityMin || h > RiskConfig.optimalHumidityMax) {
        envPoints += RiskConfig.envStressModeratePoints;
        environmentalSignals.add('humidity_out_of_range');
      }
    }

    final effectiveEnvPoints = envPoints > RiskConfig.maxEnvPoints ? RiskConfig.maxEnvPoints : envPoints;
    if (effectiveEnvPoints > 0) {
      rawScore += effectiveEnvPoints;
      structuredReasons.add(RiskReasonModel(
        code: 'environmental_stress',
        label: 'Environmental stress (Temp/Humidity)',
        value: 'Outside optimal comfort zone',
        points: effectiveEnvPoints,
        severity: 'info',
      ));
    }

    // ── 9. Recent Health Incident Signal (+5 pts) ─────────────────────────────
    if (input.recentHealthCases.isNotEmpty) {
      const pts = RiskConfig.recentCaseBonusPoints;
      rawScore += pts;
      structuredReasons.add(RiskReasonModel(
        code: 'recent_case_anomaly',
        label: 'Recent health incident reported on farm',
        value: '${input.recentHealthCases.length} recent cases in 14d',
        points: pts,
        severity: 'warning',
      ));
    }

    // ── 10. Clamp & Classify ──────────────────────────────────────────────────
    final finalScore = rawScore.clamp(0, 100);
    final riskLevel = HealthRiskLevel.fromScore(finalScore);

    final readableReasons = structuredReasons
        .map((r) => '${r.points > 0 ? "+${r.points} " : ""}${r.label} (${r.value})')
        .toList();

    return RiskAssessmentModel(
      id: '${input.caseId}_${RiskConfig.engineVersion}',
      caseId: input.caseId,
      farmId: input.farmId,
      batchId: input.batchId,
      score: finalScore,
      riskLevel: riskLevel,
      structuredReasons: structuredReasons,
      reasons: readableReasons,
      mortalityBaseline: baselineResult.baselineAverage,
      currentMortality: input.currentMortality.toDouble(),
      mortalityRatio: baselineResult.ratio,
      mortalityRate: baselineResult.mortalityRate,
      feedChangePercent: feedReduction,
      waterChangePercent: waterReduction,
      symptomSignals: symptomSignals,
      vaccinationSignals: vaccinationSignals,
      environmentalSignals: environmentalSignals,
      nearbyCaseSignal: false,
      baselineStatus: baselineResult.baselineStatus,
      dataQuality: baselineResult.dataQuality,
      inputHash: input.computeInputHash(),
      engineVersion: RiskConfig.engineVersion,
      createdAt: DateTime.now(),
      updatedAt: DateTime.now(),
    );
  }

  /// Helper to calculate historical mortality baseline and spike points
  static _MortalityAnalysisResult _analyzeMortality({
    required int currentMortality,
    required int currentBirdCount,
    required List<DailyRecordModel> recentRecords,
    required DateTime incidentDate,
  }) {
    // 1. Filter prior records (before current incident date, up to 7 latest records)
    final priorRecords = recentRecords
        .where((r) => r.recordDate.isBefore(incidentDate) || r.recordDate.isAtSameMomentAs(incidentDate))
        .take(7)
        .toList();

    final safeFlockSize = currentBirdCount > 0 ? currentBirdCount : 1000;
    final mortalityRate = currentMortality / safeFlockSize;

    // 2. Insufficient baseline fallback
    if (priorRecords.length < 2) {
      int ratePoints = 0;
      RiskReasonModel? reason;

      if (mortalityRate >= RiskConfig.mortalityRateCriticalThreshold) {
        ratePoints = RiskConfig.mortalityPointsCritical;
        reason = RiskReasonModel(
          code: 'mortality_rate_critical',
          label: 'Severe mortality rate for flock size',
          value: '${(mortalityRate * 100).toStringAsFixed(1)}% flock loss ($currentMortality birds)',
          points: ratePoints,
          severity: 'critical',
        );
      } else if (mortalityRate >= RiskConfig.mortalityRateHighThreshold) {
        ratePoints = RiskConfig.mortalityPointsHigh;
        reason = RiskReasonModel(
          code: 'mortality_rate_high',
          label: 'Elevated mortality rate for flock size',
          value: '${(mortalityRate * 100).toStringAsFixed(1)}% flock loss',
          points: ratePoints,
          severity: 'warning',
        );
      } else if (mortalityRate >= RiskConfig.mortalityRateModerateThreshold) {
        ratePoints = RiskConfig.mortalityPointsModerate;
        reason = RiskReasonModel(
          code: 'mortality_rate_moderate',
          label: 'Moderate flock mortality rate',
          value: '${(mortalityRate * 100).toStringAsFixed(1)}% flock loss',
          points: ratePoints,
          severity: 'info',
        );
      } else if (currentMortality > 0) {
        ratePoints = RiskConfig.mortalityPointsMild;
        reason = RiskReasonModel(
          code: 'mortality_limited_baseline',
          label: 'Mortality assessed using available flock data',
          value: '$currentMortality birds (Limited history)',
          points: ratePoints,
          severity: 'info',
        );
      }

      return _MortalityAnalysisResult(
        pointsAdded: ratePoints,
        baselineAverage: 0.0,
        ratio: 1.0,
        mortalityRate: mortalityRate,
        baselineStatus: 'insufficient_data',
        dataQuality: 'Limited',
        reason: reason,
      );
    }

    // 3. Robust 7-Day Baseline Calculation
    final totalPriorMortality = priorRecords.fold<int>(0, (sum, r) => sum + r.mortalityCount);
    final avgMortality = totalPriorMortality / priorRecords.length;
    final effectiveAvg = avgMortality < 1.0 ? 1.0 : avgMortality; // Prevent division anomalies
    final ratio = currentMortality / effectiveAvg;

    int points = 0;
    RiskReasonModel? reason;

    if (ratio > RiskConfig.mortalityRatioHigh) {
      points = RiskConfig.mortalityPointsCritical;
      reason = RiskReasonModel(
        code: 'mortality_spike_critical',
        label: 'Severe mortality spike anomaly',
        value: '${ratio.toStringAsFixed(1)}x baseline ($currentMortality vs ${avgMortality.toStringAsFixed(1)}/day)',
        points: points,
        severity: 'critical',
      );
    } else if (ratio > RiskConfig.mortalityRatioModerate) {
      points = RiskConfig.mortalityPointsHigh;
      reason = RiskReasonModel(
        code: 'mortality_spike_high',
        label: 'Significant mortality increase',
        value: '${ratio.toStringAsFixed(1)}x baseline',
        points: points,
        severity: 'warning',
      );
    } else if (ratio > RiskConfig.mortalityRatioMild) {
      points = RiskConfig.mortalityPointsModerate;
      reason = RiskReasonModel(
        code: 'mortality_spike_moderate',
        label: 'Mortality above historical average',
        value: '${ratio.toStringAsFixed(1)}x baseline',
        points: points,
        severity: 'warning',
      );
    } else if (ratio > RiskConfig.mortalityRatioNormal) {
      points = RiskConfig.mortalityPointsMild;
      reason = RiskReasonModel(
        code: 'mortality_spike_mild',
        label: 'Slightly elevated mortality',
        value: '${ratio.toStringAsFixed(1)}x baseline',
        points: points,
        severity: 'info',
      );
    }

    return _MortalityAnalysisResult(
      pointsAdded: points,
      baselineAverage: avgMortality,
      ratio: ratio,
      mortalityRate: mortalityRate,
      baselineStatus: 'sufficient',
      dataQuality: 'Good',
      reason: reason,
    );
  }
}

class _MortalityAnalysisResult {
  final int pointsAdded;
  final double baselineAverage;
  final double ratio;
  final double mortalityRate;
  final String baselineStatus;
  final String dataQuality;
  final RiskReasonModel? reason;

  const _MortalityAnalysisResult({
    required this.pointsAdded,
    required this.baselineAverage,
    required this.ratio,
    required this.mortalityRate,
    required this.baselineStatus,
    required this.dataQuality,
    this.reason,
  });
}

File: functions/src/health/riskEngine.ts
import * as crypto from 'crypto';
import { RiskConfig } from './riskConfig';

export interface RiskReason {
  code: string;
  label: string;
  value: string;
  points: number;
  severity: 'info' | 'warning' | 'critical';
}

export interface RiskEngineInput {
  caseId: string;
  farmId: string;
  batchId: string;
  currentMortality: number;
  affectedCount: number;
  symptoms: string[];
  feedReductionPercent?: number;
  waterReductionPercent?: number;
  temperature?: number;
  humidity?: number;
  incidentDate: Date;
  currentBirdCount?: number;
  recentDailyMortalities?: number[];
  hasOverdueVaccine?: boolean;
  hasVaccineRecords?: boolean;
  recentHealthCasesCount?: number;
}

export interface RiskAssessmentResult {
  id: string;
  caseId: string;
  farmId: string;
  batchId: string;
  score: number;
  riskLevel: 'low' | 'moderate' | 'high' | 'critical';
  structuredReasons: RiskReason[];
  reasons: string[];
  mortalityBaseline: number;
  currentMortality: number;
  mortalityRatio: number;
  mortalityRate: number;
  feedChangePercent: number;
  waterChangePercent: number;
  symptomSignals: string[];
  vaccinationSignals: string[];
  environmentalSignals: string[];
  baselineStatus: string;
  dataQuality: string;
  inputHash: string;
  engineVersion: string;
  createdAt: Date;
  updatedAt: Date;
}

export function computeInputHash(input: RiskEngineInput): string {
  const sortedSymptoms = [...input.symptoms].sort().join(',');
  const raw = [
    input.caseId,
    input.farmId,
    input.batchId,
    input.currentMortality,
    input.affectedCount,
    sortedSymptoms,
    input.feedReductionPercent?.toFixed(1) ?? '0',
    input.waterReductionPercent?.toFixed(1) ?? '0',
    input.temperature?.toFixed(1) ?? '0',
    input.humidity?.toFixed(1) ?? '0',
    input.hasOverdueVaccine ? '1' : '0',
    RiskConfig.engineVersion,
  ].join('|');

  return crypto.createHash('md5').update(raw).digest('hex');
}

export function calculateHealthRisk(input: RiskEngineInput): RiskAssessmentResult {
  const structuredReasons: RiskReason[] = [];
  const symptomSignals: string[] = [];
  const vaccinationSignals: string[] = [];
  const environmentalSignals: string[] = [];
  let rawScore = 0;

  // 1. Mortality Baseline & Ratio
  const recent = input.recentDailyMortalities ?? [];
  const flockSize = (input.currentBirdCount && input.currentBirdCount > 0) ? input.currentBirdCount : 1000;
  const mortalityRate = input.currentMortality / flockSize;
  let baselineAvg = 0;
  let ratio = 1.0;
  let baselineStatus = 'sufficient';
  let dataQuality = 'Good';

  if (recent.length < 2) {
    baselineStatus = 'insufficient_data';
    dataQuality = 'Limited';
    let pts = 0;
    let severity: 'info' | 'warning' | 'critical' = 'info';

    if (mortalityRate >= RiskConfig.mortalityRateCriticalThreshold) {
      pts = RiskConfig.mortalityPointsCritical;
      severity = 'critical';
    } else if (mortalityRate >= RiskConfig.mortalityRateHighThreshold) {
      pts = RiskConfig.mortalityPointsHigh;
      severity = 'warning';
    } else if (mortalityRate >= RiskConfig.mortalityRateModerateThreshold) {
      pts = RiskConfig.mortalityPointsModerate;
      severity = 'info';
    } else if (input.currentMortality > 0) {
      pts = RiskConfig.mortalityPointsMild;
    }

    if (pts > 0) {
      rawScore += pts;
      structuredReasons.push({
        code: 'mortality_limited_baseline',
        label: 'Mortality assessed using flock population rate',
        value: `${(mortalityRate * 100).toFixed(1)}% flock loss (${input.currentMortality} birds)`,
        points: pts,
        severity,
      });
    }
  } else {
    const sum = recent.reduce((a, b) => a + b, 0);
    baselineAvg = sum / recent.length;
    const effectiveAvg = baselineAvg < 1.0 ? 1.0 : baselineAvg;
    ratio = input.currentMortality / effectiveAvg;

    let pts = 0;
    let severity: 'info' | 'warning' | 'critical' = 'info';
    let label = 'Slightly elevated mortality';

    if (ratio > RiskConfig.mortalityRatioHigh) {
      pts = RiskConfig.mortalityPointsCritical;
      severity = 'critical';
      label = 'Severe mortality spike anomaly';
    } else if (ratio > RiskConfig.mortalityRatioModerate) {
      pts = RiskConfig.mortalityPointsHigh;
      severity = 'warning';
      label = 'Significant mortality increase';
    } else if (ratio > RiskConfig.mortalityRatioMild) {
      pts = RiskConfig.mortalityPointsModerate;
      severity = 'warning';
      label = 'Mortality above historical average';
    } else if (ratio > RiskConfig.mortalityRatioNormal) {
      pts = RiskConfig.mortalityPointsMild;
      severity = 'info';
    }

    if (pts > 0) {
      rawScore += pts;
      structuredReasons.push({
        code: 'mortality_spike',
        label,
        value: `${ratio.toFixed(1)}x baseline (${input.currentMortality} vs ${baselineAvg.toFixed(1)}/day)`,
        points: pts,
        severity,
      });
    }
  }

  // 2. Feed Reduction
  const feed = input.feedReductionPercent ?? 0;
  if (feed > RiskConfig.feedReductionCritical) {
    rawScore += RiskConfig.feedPointsCritical;
    structuredReasons.push({
      code: 'feed_reduction_critical',
      label: 'Severe feed consumption drop',
      value: `-${feed.toFixed(0)}% intake loss`,
      points: RiskConfig.feedPointsCritical,
      severity: 'critical',
    });
  } else if (feed > RiskConfig.feedReductionHigh) {
    rawScore += RiskConfig.feedPointsHigh;
    structuredReasons.push({
      code: 'feed_reduction_high',
      label: 'Significant feed intake reduction',
      value: `-${feed.toFixed(0)}%`,
      points: RiskConfig.feedPointsHigh,
      severity: 'warning',
    });
  } else if (feed > RiskConfig.feedReductionModerate) {
    rawScore += RiskConfig.feedPointsModerate;
    structuredReasons.push({
      code: 'feed_reduction_moderate',
      label: 'Moderate feed intake drop',
      value: `-${feed.toFixed(0)}%`,
      points: RiskConfig.feedPointsModerate,
      severity: 'warning',
    });
  } else if (feed > RiskConfig.feedReductionMild) {
    rawScore += RiskConfig.feedPointsMild;
    structuredReasons.push({
      code: 'feed_reduction_mild',
      label: 'Minor feed intake reduction',
      value: `-${feed.toFixed(0)}%`,
      points: RiskConfig.feedPointsMild,
      severity: 'info',
    });
  }

  // 3. Water Reduction
  const water = input.waterReductionPercent ?? 0;
  if (water > RiskConfig.waterReductionCritical) {
    rawScore += RiskConfig.waterPointsCritical;
    structuredReasons.push({
      code: 'water_reduction_critical',
      label: 'Severe water consumption drop',
      value: `-${water.toFixed(0)}% water loss`,
      points: RiskConfig.waterPointsCritical,
      severity: 'critical',
    });
  } else if (water > RiskConfig.waterReductionHigh) {
    rawScore += RiskConfig.waterPointsHigh;
    structuredReasons.push({
      code: 'water_reduction_high',
      label: 'Significant water intake drop',
      value: `-${water.toFixed(0)}%`,
      points: RiskConfig.waterPointsHigh,
      severity: 'warning',
    });
  } else if (water > RiskConfig.waterReductionModerate) {
    rawScore += RiskConfig.waterPointsModerate;
    structuredReasons.push({
      code: 'water_reduction_moderate',
      label: 'Moderate water intake drop',
      value: `-${water.toFixed(0)}%`,
      points: RiskConfig.waterPointsModerate,
      severity: 'warning',
    });
  } else if (water > RiskConfig.waterReductionMild) {
    rawScore += RiskConfig.waterPointsMild;
    structuredReasons.push({
      code: 'water_reduction_mild',
      label: 'Minor water intake reduction',
      value: `-${water.toFixed(0)}%`,
      points: RiskConfig.waterPointsMild,
      severity: 'info',
    });
  }

  // 4. Symptoms
  let rawSymptomPts = 0;
  const normalizedSymptoms = input.symptoms.map(s => s.trim().toLowerCase().replace(/ /g, '_'));
  for (const sym of normalizedSymptoms) {
    const weight = RiskConfig.symptomWeights[sym] ?? 4;
    rawSymptomPts += weight;
    symptomSignals.push(sym);
  }
  const effectiveSymptomPts = Math.min(rawSymptomPts, RiskConfig.maxSymptomPoints);
  if (effectiveSymptomPts > 0) {
    rawScore += effectiveSymptomPts;
    structuredReasons.push({
      code: 'clinical_symptoms',
      label: 'Clinical symptoms reported',
      value: `${normalizedSymptoms.length} signs (${effectiveSymptomPts} pts)`,
      points: effectiveSymptomPts,
      severity: effectiveSymptomPts >= 15 ? 'critical' : (effectiveSymptomPts >= 8 ? 'warning' : 'info'),
    });
  }

  // 5. Syndromic Clusters
  const respMatches = normalizedSymptoms.filter(s => RiskConfig.respiratorySymptoms.includes(s)).length;
  if (respMatches >= 2) {
    rawScore += RiskConfig.syndromeBonusPoints;
    structuredReasons.push({
      code: 'respiratory_syndrome',
      label: 'Respiratory syndrome detected',
      value: `${respMatches} distinct respiratory signs`,
      points: RiskConfig.syndromeBonusPoints,
      severity: 'critical',
    });
  }

  // 6. Sudden Death
  if (normalizedSymptoms.includes('sudden_death')) {
    rawScore += RiskConfig.suddenDeathBonusPoints;
    structuredReasons.push({
      code: 'sudden_death_event',
      label: 'Sudden acute mortality event reported',
      value: 'Peracute onset',
      points: RiskConfig.suddenDeathBonusPoints,
      severity: 'critical',
    });
  }

  // 7. Vaccination
  if (input.hasOverdueVaccine) {
    rawScore += RiskConfig.vaccineSignificantlyOverduePoints;
    vaccinationSignals.push('overdue');
    structuredReasons.push({
      code: 'vaccine_overdue',
      label: 'Flock immunization overdue',
      value: 'Missing scheduled vaccine',
      points: RiskConfig.vaccineSignificantlyOverduePoints,
      severity: 'warning',
    });
  } else if (!input.hasVaccineRecords) {
    rawScore += RiskConfig.vaccineUnknownPoints;
    vaccinationSignals.push('unknown');
    structuredReasons.push({
      code: 'vaccine_unverified',
      label: 'Vaccination records unrecorded',
      value: 'Incomplete log',
      points: RiskConfig.vaccineUnknownPoints,
      severity: 'info',
    });
  }

  // 8. Environmental Stress
  let envPts = 0;
  if (input.temperature && (input.temperature < RiskConfig.optimalTempMin || input.temperature > RiskConfig.optimalTempMax)) {
    envPts += RiskConfig.envStressModeratePoints;
    environmentalSignals.push('temp_out_of_range');
  }
  if (input.humidity && (input.humidity < RiskConfig.optimalHumidityMin || input.humidity > RiskConfig.optimalHumidityMax)) {
    envPts += RiskConfig.envStressModeratePoints;
    environmentalSignals.push('humidity_out_of_range');
  }
  const effectiveEnvPts = Math.min(envPts, RiskConfig.maxEnvPoints);
  if (effectiveEnvPts > 0) {
    rawScore += effectiveEnvPts;
    structuredReasons.push({
      code: 'environmental_stress',
      label: 'Environmental stress (Temp/Humidity)',
      value: 'Outside optimal comfort zone',
      points: effectiveEnvPts,
      severity: 'info',
    });
  }

  // 9. Recent Cases
  if (input.recentHealthCasesCount && input.recentHealthCasesCount > 0) {
    rawScore += RiskConfig.recentCaseBonusPoints;
    structuredReasons.push({
      code: 'recent_case_anomaly',
      label: 'Recent health incident reported on farm',
      value: `${input.recentHealthCasesCount} active cases in 14d`,
      points: RiskConfig.recentCaseBonusPoints,
      severity: 'warning',
    });
  }

  // 10. Clamp & Classify
  const finalScore = Math.max(0, Math.min(rawScore, 100));
  let riskLevel: 'low' | 'moderate' | 'high' | 'critical' = 'low';
  if (finalScore >= 76) riskLevel = 'critical';
  else if (finalScore >= 51) riskLevel = 'high';
  else if (finalScore >= 31) riskLevel = 'moderate';

  const readableReasons = structuredReasons.map(
    r => `${r.points > 0 ? '+' + r.points + ' ' : ''}${r.label} (${r.value})`
  );

  return {
    id: `${input.caseId}_${RiskConfig.engineVersion}`,
    caseId: input.caseId,
    farmId: input.farmId,
    batchId: input.batchId,
    score: finalScore,
    riskLevel,
    structuredReasons,
    reasons: readableReasons,
    mortalityBaseline: baselineAvg,
    currentMortality: input.currentMortality,
    mortalityRatio: ratio,
    mortalityRate,
    feedChangePercent: feed,
    waterChangePercent: water,
    symptomSignals,
    vaccinationSignals,
    environmentalSignals,
    baselineStatus,
    dataQuality,
    inputHash: computeInputHash(input),
    engineVersion: RiskConfig.engineVersion,
    createdAt: new Date(),
    updatedAt: new Date(),
  };
}
